src/transport/manager/mod.rs

```diff
@@ -3266,4 +3357,232 @@ mod tests {
         }
     }
+
+    #[tokio::test]
+    async fn do_not_overwrite_dial_addresses() {
+        let _ = tracing_subscriber::fmt()
+            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+            .try_init();
+
+        let (mut manager, _handle) = TransportManager::new(
+            Keypair::generate(),
+            HashSet::new(),
+            BandwidthSink::new(),
+            8usize,
+        );
+        let peer = PeerId::random();
+        let dial_address = Multiaddr::empty()
+            .with(Protocol::Ip4(Ipv4Addr::new(127, 0, 0, 1)))
+            .with(Protocol::Tcp(8888))
+            .with(Protocol::P2p(
+                Multihash::from_bytes(&peer.to_bytes()).unwrap(),
+            ));
+
+        let connection_id = ConnectionId::from(0);
+        let transport = Box::new({
+            let mut transport = DummyTransport::new();
+            transport.inject_event(TransportEvent::ConnectionEstablished {
+                peer,
+                endpoint: Endpoint::listener(dial_address.clone(), connection_id),
+            });
+            transport
+        });
+        manager.register_transport(SupportedTransport::Tcp, transport);
+
+        // First dial attempt.
+        manager.dial_address(dial_address.clone()).await.unwrap();
+        // check the state of the peer.
+        {
+            let peers = manager.peers.read();
+            let peer_context = peers.get(&peer).unwrap();
+            match &peer_context.state {
+                PeerState::Dialing { record } => {
+                    assert_eq!(record.address(), &dial_address);
+                }
+                state => panic!("invalid state: {state:?}"),
+            }
+
+            // The address is not saved yet.
+            assert!(!peer_context.addresses.contains(&dial_address));
+        }
+
+        let second_address = Multiaddr::empty()
+            .with(Protocol::Ip4(Ipv4Addr::new(127, 0, 0, 1)))
+            .with(Protocol::Tcp(8889))
+            .with(Protocol::P2p(
+                Multihash::from_bytes(&peer.to_bytes()).unwrap(),
+            ));
+
+        // Second dial attempt with different address.
+        manager.dial_address(second_address.clone()).await.unwrap();
+        // check the state of the peer.
+        {
+            let peers = manager.peers.read();
+            let peer_context = peers.get(&peer).unwrap();
+            match &peer_context.state {
+                // Must still be dialing the first address.
+                PeerState::Dialing { record } => {
+                    assert_eq!(record.address(), &dial_address);
+                }
+                state => panic!("invalid state: {state:?}"),
+            }
+
+            assert!(!peer_context.addresses.contains(&dial_address));
+            assert!(!peer_context.addresses.contains(&second_address));
+        }
+    }
+
+    #[tokio::test]
+    async fn reject_unknown_secondary_connections_with_different_connection_ids() {
```

Comment:
How do we check for the unknown connection id being rejected below? Or we don't have the issue of getting the unknown connection id after this PR?

Reply:
Indeed the edge-case that goes from those states is fixed now, I've added another test to check we reject connections from a theoretical state.

We might still have some other gaps, the state machine is quite complex